caffe2/serialize/intermediate_model.cc

```diff
@@ -0,0 +1,334 @@
+          // TODO
+          AT_ERROR("Storing data in separate file is not supported yet!");
+        } else {
+          // TODO
+          AT_ERROR("Unknown source_type: ", caffe2::to_string(source_type));
+        }
+        break;
+      }
+    case caffe2::TensorProto_StorageType_NO_CONTENT:
+      {
+        noContent_ = true;
+        break;
+      }
+    default:
+      AT_ERROR("Uknown storage_type: ", caffe2::to_string(storage_type));
+  }
+
+}
+
+void IntermediateTensor::dump(caffe2::TensorProto* tensor_proto) {
+  for (auto dim : dims_) {
+    tensor_proto->add_dims(dim);
+  }
+  tensor_proto->set_data_type(static_cast<caffe2::TensorProto_DataType>(dataType_));
+  // NB: maybe later we support RAW
+  tensor_proto->set_storage_type(caffe2::TensorProto_StorageType::TensorProto_StorageType_EXTERNAL);
+  caffe2::ExternalDataProto* data_proto = tensor_proto->mutable_external_data();
+  // NB: maybe later we support SIMPLE_FILE
+  data_proto->set_source_type(caffe2::ExternalDataProto_SourceType_INLINE_CONTAINER);
+  AT_ASSERTM(data_->recordId.has_value(), "recordId is required for SharedData!");
+  data_proto->set_record_id(caffe2::to_string(data_->recordId.value()));
+  data_proto->set_offset(offset_);
+  for (auto stride : strides_) {
+    data_proto->add_strides(stride);
+  }
+  caffe2::DeviceOption* device_detail = tensor_proto->mutable_device_detail();
+  device_detail->set_device_type(deviceDetail_.deviceType);
+  if (deviceDetail_.deviceId.has_value()) {
+    device_detail->set_device_id(deviceDetail_.deviceId.value());
+  }
+}
+
+IntermediateParameter::IntermediateParameter(torch::ParameterDef* param_def,
+    std::unordered_map<uint64_t, std::shared_ptr<SharedData>>* id_data,
+    DeserializeMode mode) {
+  AT_ASSERTM(param_def->has_name(), "ParameterDef has no name! ",
+      param_def->DebugString());
+  name_ = param_def->name();
+  isBuffer_ = param_def->is_buffer();
+  requireGradient_ = param_def->require_gradient();
+  if (param_def->has_tensor()) {
+    tensor_.update(param_def->mutable_tensor(), id_data, mode);
+  } else {
+    // TODO
+    AT_ERROR("A ParameterDef does not contain any tensor!");
+  }
+}
+
+void IntermediateParameter::dump(torch::ParameterDef* param_def) {
+  param_def->set_name(name_);
+  param_def->set_is_buffer(isBuffer_);
+  param_def->set_require_gradient(requireGradient_);
+  caffe2::TensorProto* tensor_def = param_def->mutable_tensor();
+  tensor_.dump(tensor_def);
+}
+
+IntermediateMethod::IntermediateMethod(torch::MethodDef* method_def) {
+  AT_ASSERTM(method_def->has_name(), "name is required for MethodDef!");
+  name_ = method_def->name();
+  if (method_def->has_torch_script()) {
+    torchScript_ = method_def->torch_script();
+  } else if (method_def->has_graph()) {
+    graph_.reset(method_def->release_graph());
+  } else {
+    AT_ERROR("No method body is found!");
+  }
+}
+
+IntermediateMethod::IntermediateMethod(const IntermediateMethod& method) {
```

Comment:
why do you ever need to have copy / assignment constructors for these?

Reply:
because windows build ask for `IntermediateMethod& IntermediateMethod::operator =(const IntermediateMethod& method)`, and if I add it, linux started asking for `IntermediateMethod::IntermediateMethod(const IntermediateMethod& method) `

Reply:
I think by default it's deleted on Windows

Reply:
wait, why does windows build require a copy constructor? it's probably because you stash things into a vector. So you could add a move constructor (marked noexcept) and then delete the copy constructor. (or the lazy solution - store a vector of unique pointers)

There shouldn't be a place where you copy the IntermediateMethod, really.